compiler/src/dotty/tools/dotc/semanticdb/SymbolOps.scala

```diff
@@ -0,0 +1,218 @@
+          s.SuperType(spre, ssym)
+
+        // val clazzOf = classOf[...]
+        case ConstantType(const) if const.tag == core.Constants.ClazzTag =>
+          loop(const.typeValue)
+
+        case ConstantType(const) =>
+          s.ConstantType(const.toSemanticConst)
+
+        case rt @ RefinedType(parent, name, info) =>
+          // `X { def x: Int; def y: Int }`
+          //
+          // RefinedType(
+          //   parent = RefinedType(
+          //     parent = TypeRef(..., X)
+          //     ...
+          //   )
+          //   refinedName = x
+          //   refinedInfo = TypeRef(..., Int)
+          // )
+          def getParent(parent: Type): Type = parent match {
+            case RefinedType(p, _, _) => getParent(p)
+            case _ => parent
+          }
+
+          val realParent = getParent(parent)
+          val stpe = realParent match {
+            // val tp: M with N { def k: Int } = ???
+            case AndType(x, y) =>
+              s.WithType(Seq(loop(x), loop(y))) // TODO: for M with N with L
+            case _ =>
+              s.WithType(Seq(loop(parent)))
+          }
+          // TODO: collect all refinement symbols
+          // val sdecls = decls.sscope(using LinkMode.HardlinkChildren)
+          s.StructuralType(stpe, None)
+
+        case rec: RecType =>
+          loop(rec.parent) // should be handled as RefinedType
+
+        // repeated params: e.g. `Int*`, which is the syntax sugar of
+        // `Seq[Int] @Repeated` (or `Array[Int] @Repeated`)
+        // See: Desugar.scala and TypeApplications.scala
+        case AnnotatedType(AppliedType(_, targs), annot)
+          if (annot matches defn.RepeatedAnnot) && (targs.length == 1) =>
+          val stpe = loop(targs(0))
+          s.RepeatedType(stpe)
+
+        case ann: AnnotatedType if ann.annot.symbol.info.isInstanceOf[ClassInfo] =>
+          def flatten(tpe: Type, annots: List[Annotation]): (Type, List[Annotation]) = tpe match
+            case AnnotatedType(parent, annot) if annot.symbol.info.isInstanceOf[ClassInfo] =>
+              flatten(parent, annot +: annots)
+            case other => (other, annots)
+
+          val (parent, annots) = flatten(ann, List.empty)
+          val sparent = loop(parent)
+          val sannots = annots.map(a =>
+            s.Annotation(loop(a.symbol.info.asInstanceOf[ClassInfo].selfType))
+          )
+          s.AnnotatedType(sannots, sparent)
+
+        case app @ AppliedType(tycon, args) =>
+          val sargs = args.map(loop)
+          loop(tycon) match
+            case ref: s.TypeRef => ref.copy(typeArguments = sargs)
+            // is there any other cases?
+            case _ => s.Type.Empty
+
+        case and: AndType =>
+          def flatten(child: Type): List[Type] = child match
+            case AndType(ct1, ct2) => flatten(ct1) ++ flatten(ct2)
+            case other => List(other)
+          val stpes = flatten(and).map(loop)
+          s.IntersectionType(stpes)
```

Comment:
Could you elaborate on it?
I thought modeling [Scala3's IntersectionType](https://dotty.epfl.ch/docs/reference/new-types/intersection-types.html) by [Semanticdb's IntersetionType](https://scalameta.org/docs/semanticdb/specification.html#type) seems legid, and while `IntersectionType` is commutative `WithType` has a sense of non-commutative.

Reply:
So with semanticDB's extraction we want to fit Scala 3 definitions within semanticDB's model of the world, semanticdb has no knowledge of Scala 3's intersection types, so we map it to the next best thing - Scala 2's compound types, (i.e. `WithType`). We already make some "sacrifices" in this way, e.g. there is no `given` flag in semanticdb, so we use `implicit` instead.

In the future we probably want to add more Scala 3 specialised definitions to the protobuf schema.

Reply:
Hmm, sorry I still can't see the reason for using `WithType` here, I might be missing something...
What do you think about this? @tgodzik 

> semanticDB has no knowledge of Scala 3's intersection types
- In the sense of SemanticDB supports the model that can represent Scala3's intersection type, SemanticDB already supports `IntersectionType`.
  - I think things are different from the case of `given`, while SemanticDB doesn't have an appropriate property for `given` and we have to "sacrifice" (or "erase"?) the property down to `implicit`, we already have `IntersectionType` in Semanticdb and it has "equivalent" with Scala3's IntersetionType
- If you mean the Semanticdb generated from Scala3 should be compatible with Scala2
  - I don't think we need the compatibility, because unlike tasty, Semanticdb is basically generated and consumed for the same version of Scala.
  - what about `UnionType`?
  - I'm planning to add some Scala3 specialized types to Semanticdb protobuf schema for `TypeLambda` and `MatchType` because I think current SemanticDB's Types are incapable of representing those Scala3 types.
    - Should we also somehow encode them with Scala2 compatible type?

Reply:
It seems `UnionType` is never used, so it is free to interpret how you wish - My comment about `IntersectionType` was in case some tooling makes an assumption that it implies the definition must be a Java type, but maybe this is not the case in practise. It is true that a Java intersection type has different erasure to both a scala 3 intersection type or a Scala 2 `WithType`, so it might be worth treating separately

Reply:
Ah! I got it, so the problem is the interpretation of `IntersectionType` could be ambiguous whether it is for Java methods or Intersection Type in Scala3.
Indeed it could be better to use `WithType` instead... let me sleep on it 🤔